packages/flutter/lib/src/widgets/visibility.dart

```diff
@@ -0,0 +1,209 @@
+       assert(maintainSize == true || maintainSemantics == false, 'Cannot maintain semantics if size is not maintained.'),
+       assert(maintainSize == true || maintainInteractivity == false, 'Cannot maintain interactivity if size is not maintained.'),
+       super(key: key);
+
+  /// The widget to show or hide, as controlled by [visible].
+  ///
+  /// {@macro flutter.widgets.child}
+  final Widget child;
+
+  /// The widget to use when the child is not [visible], assuming that none of
+  /// the `maintain` flags (in particular, [maintainState]) are set.
+  ///
+  /// The normal behavior is to replace the widget with a zero by zero box
+  /// ([SizedBox.shrink]).
+  ///
+  /// See also:
+  ///
+  ///  * [AnimatedCrossFade], which can animate between two children.
+  final Widget replacement;
+
+  /// Switches between showing the [child] or hiding it.
+  ///
+  /// The `maintain` flags should be set to the same values regardless of the
+  /// state of the [visible] property, otherwise they will not operate correctly
+  /// (specifically, the state will be lost regardless of the state of
+  /// [maintainState] whenever any of the `maintain` flags are changed, since
+  /// doing so will result in a subtree shape change).
+  final bool visible;
+
+  /// Whether to maintain the [State] objects of the [child] subtree when it is
+  /// not [visible].
+  ///
+  /// Keeping the state of the subtree is expensive. It should only be
+  /// maintained if it cannot be recreated on demand. One example of when the
+  /// state would be maintained is if the child subtree contains a [Navigator],
+  /// since that widget maintains elaborate state that cannot be recreated on
+  /// the fly.
+  ///
+  /// If this is true, an [Offstage] widget is used to hide the child instead of
+  /// replacing it with [replacement].
+  final bool maintainState;
+
+  /// Whether to maintain animations within the [child] subtree when it is
+  /// not [visible].
+  ///
+  /// To set this, [maintainState] must also be set.
+  ///
+  /// Keeping animations active when the widget is not visible is even more
+  /// expensive than only maintaining the state.
+  ///
+  /// One example when this might be useful is if the subtree contains an
+  /// animated [Image] that must be kept synchronized with other animated images
+  /// elsewhere in the interface.
+  ///
+  /// If this is true, no [TickerMode] widget is used.
+  final bool maintainAnimation;
+
+  /// Whether to maintain space for where the widget would have been.
+  ///
+  /// To set this, [maintainAnimation] must also be set.
+  ///
+  /// Maintaining the size when the widget is not [visible] is roughly as
+  /// expensive as keeping animations running, and may in some circumstances be
+  /// cheaper if the subtree is simple and the [visible] property is frequently
+  /// toggled, since it avoids triggering a layout change when the [visible]
```

Comment:
It's not clear after reading this sentence what's the choice the developer needs to make or that maintainSize implies maintainAnimation.

Reply:
reworded to make this clearer.